packages/math/base-elements.lua

```diff
@@ -0,0 +1,1209 @@
+    -- an advance measurement as close as possible as the required one.
+    -- The advance measurement is simply the depth+height of the glyph.
+    -- Therefore, the selected glyph may be smaller or bigger than
+    -- required.  TODO: implement assembly of stretchable glyphs form
+    -- their parts for cases when the biggest variant is not big enough.
+    -- We copy the glyph list to avoid modifying the shaper's cache. Yes.
+    local glyphs = std.tree.clone(self.value.items)
+    local constructions = getMathMetrics().mathVariants
+      .vertGlyphConstructions[glyphs[1].gid]
+    if constructions then
+      local variants = constructions.mathGlyphVariantRecord
+      SU.debug("math", "stretch: variants = " .. variants)
+      local closest
+      local closestI
+      local m = requiredAdvance - (self.depth+self.height):tonumber() * upem/sz
+      SU.debug("math", "strech: m = " .. m)
+      for i,v in ipairs(variants) do
+        local diff = math.abs(v.advanceMeasurement - requiredAdvance)
+        SU.debug("math", "stretch: diff = " .. diff)
+        if diff < m then
+          closest = v
+          closestI = i
+          m = diff
+        end
+      end
+      SU.debug("math", "stretch: closestI = " .. closestI)
+      if closest then
+        -- Now we have to re-shape the glyph chain. We will assume there
+        -- is only one glyph.
+        -- TODO: this code is probably wrong when the vertical
+        -- variants have a different width than the original, because
+        -- the shaping phase is already done. Need to do better.
+        glyphs[1].gid = closest.variantGlyph
+        local face = SILE.font.cache(self.options, SILE.shaper.getFace)
+        local dimen = hb.get_glyph_dimensions(face.data,
+          face.index, self.options.size, closest.variantGlyph)
+        glyphs[1].width = dimen.width
+        glyphs[1].height = dimen.height
+        glyphs[1].depth = dimen.depth
+        glyphs[1].glyphAdvance = dimen.glyphAdvance
+        self.width = SILE.length(dimen.glyphAdvance)
+        self.depth = SILE.length(dimen.depth)
+        self.height = SILE.length(dimen.height)
+        SILE.shaper:preAddNodes(glyphs, self.value)
+        self.value.items = glyphs
+        self.value.glyphString = {glyphs[1].gid}
+      end
+    end
+  end,
+
+  output = function(self, x, y, line)
+    if not self.value.glyphString then return end
+    local compensatedY
+    if isDisplayMode(self.mode) and self.atom == atomType.bigOperator
+        and self.value.items[1].fontDepth then
+      compensatedY = SILE.length(y.length + self.value.items[1].depth
+        - self.value.items[1].fontDepth)
+    else
+      compensatedY = y
+    end
+    SILE.outputter:setCursor(scaleWidth(x, line), compensatedY.length)
+    SILE.outputter:setFont(self.options)
+    pdf.colorpush_rgb(0,0,0)
+    pdf.colorpop()
+    SILE.outputter:drawHbox(self.value, scaleWidth(self.width, line))
```

Comment:
Related to #1121, why does the second argument to `drawHbox()` here have all this fancy math? In my testing setting this to and integer at all shows no difference in rendering.

Reply:
You’re right, in the case of a text element, the width has no reason to be stretchable or shrinkable.

Reply:
It's not just that you don't need to do scaling here, it's that the value doesn't seem to be used for anything at all:

```lua
SILE.outputter:drawHbox(self.value, 0)
SILE.outputter:drawHbox(self.value, 999)
```

Reply:
Well, I’m not the one who wrote `SILE.outputter:drawHbox`, I must admit I did try to understand why it takes the width as an argument. I my latest commits I removed the unnecessary `scaleWidth`, but I would need some more time to understand why the width is here in the first place.